src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs

```diff
@@ -65,38 +78,79 @@ private static void Scan(object state)
 
         private void Scan()
         {
-            // Scan the registered connections looking for ones that have timed out
-            foreach (var c in _connections)
+            if (_running)
             {
-                if (!c.Value.Active && (DateTimeOffset.UtcNow - c.Value.LastSeenUtc).TotalSeconds > 5)
+                return;
+            }
+
+            try
+            {
+                _running = true;
+
+                // Scan the registered connections looking for ones that have timed out
+                foreach (var c in _connections)
                 {
-                    ConnectionState s;
-                    if (_connections.TryRemove(c.Key, out s))
+                    var status = ConnectionState.ConnectionStatus.Inactive;
+                    var lastSeenUtc = DateTimeOffset.UtcNow;
+
+                    try
+                    {
+                        c.Value.Lock.Wait();
+
+                        // Capture the connection state
+                        status = c.Value.Status;
+
+                        lastSeenUtc = c.Value.LastSeenUtc;
+                    }
+                    finally
                     {
-                        // REVIEW: Should we keep firing and forgetting this?
-                        var ignore = s.DisposeAsync();
+                        c.Value.Lock.Release();
+                    }
+
+                    // Once the decision has been made to to dispose we don't check the status again
+                    if (status == ConnectionState.ConnectionStatus.Inactive && (DateTimeOffset.UtcNow - lastSeenUtc).TotalSeconds > 5)
```

Comment:
There's a race here, but I suppose we don't have to get too worked up about it... It just feels odd to be adding a race just to save a few microseconds inside the lock. We can make the determination that we're going to dispose and change the state to "Disposing" or something inside the lock and still do the actual Dispose outside the lock. That prevents the race.